### PR TITLE
Handle Disconnection

### PR DESCRIPTION
**Pull request by flawedmatrix**
CHANGES
- Added Game controller method for removing players from the player list
- Refactored socket communication so that game_id, player_id, isHost, and map_id is generated onJoin, which is possible because the user is stored server side in the initial socket.io request
- This means all of the hacky game.js code dealing with game_id and player_id and storing temporary variables in divs in the HTML is redundant, so the code has been removed and tests changed to reflect the simpler interface
- This was necessary to make handling leave_room logic very simple
- Wrote tests in Jasmine, for game.js leave logic, wrote test for game controller method for removing players
- Wrote tests for Socket.io not only for leave room logic but for everything else
- Guests can no longer start the game by sending the 'start' packet
- Added mock library for Python (makes Mock testing easy)
- Added extra loading screen to prevent the start button from being visible until finishInitialization()

Here is how socket communication works now:
1. Client GET /game/{ gameid }
2. Server processes link, renders { gameid } back to client through the template and directly passing it into the Game() constructor
3. On Game.init() the client makes a socket.io request with the /game namespace and the 'id={ gameid }' query string
4. Server handles the initial connection, retrieving the request.user object (django magically handles the sessions and cookies to do this) and parsing the query string
5. Client then sends the 'join' message indicating the client wants to join the game
6. Server automatically adds the client to the '{ gameid }' room and adds the user to the game in the database. It then sends back a 'game_data' packet containing player_id, isHost, map_id, and player_list, which all gets used in the client.

tl;dr:
Loading screen
User disconnects/leaving room logic
Don't send the server the game_id and player_id; it already knows what they are.
Make sure you do pip install again

### CODE REVIEW

File: game/sockets.py [hunk 1]
```diff
@@ -37,19 +46,57 @@ def on_click(self, data):
...
+
+        self.session['player_id'] = player_id
 
-    def on_join(self, data):
-        self.join(str(data['game_id']))
-        self.broadcast_message('join', data)
+        # The host of the game is the person that created the game
+        # WHICH HAPPENS TO BE player_id == 0 IN OUR IMPLEMENTATION
+        # Perhaps we should fix this one day
+        isHost = self.session['isHost'] = (player_id == 0)
```
Comment (hellochar) [inline game/sockets.py:73]: Are you sure this was always the case? The Game model has a "creator" field

Comment (flawedmatrix) [inline game/sockets.py:73]: It was in there at least. We could always change this. Another problem there was was that games can be emptied while in the lobby state and the game would stay available. Then the first person who entered the empty map would be able to become host.
It would be a good idea to allow the host to change instead of fixing it to the game creator. For now this is a decent way to fix the host to only one person and if people wanted to change they host they could both just leave the page and have the other go in first.
Of course right now there really is no advantage of having a host other than the fact that they have the start button, but later down the road they could make the game more interesting.
I feel like on order to fully flesh out the isHost part for the future we need to do some more designing.

**Comment by hellochar** [comment]
looks good; did you test the disconnect/leave logic? what does it look like?

**Comment by flawedmatrix** [comment]
It looks how I think it should:
If you leave, everyone sees that you're gone from the list.

Also everything should have been tested (I checked visually and by code).

File: game/static/game/js/game.js [hunk 2]
```diff
@@ -206,69 +197,47 @@ Game.prototype.renderMethod = function() {
...
+    }
+    if (this.keys.a) {
+        delta.x -= 40;
+    }
+    if (this.keys.s) {
+        delta.y -= 40;
     }
+    if (this.keys.d) {
+        delta.x += 40;
+    }
```
Comment (jamesmuerle) [inline game/static/game/js/game.js:228]: Extract this '40' to a configuration variable. It should be stored in the game and initialized when a client receives the packet containing the Game information. In the future, this value will be user-dependent (a gameplay option).

**Comment by jamesmuerle** [comment]
Looks good overall.

Edwin, for all your future programming endeavors, this is the best advice I can give you:

Communicate what your changes are if they extend beyond what is obvious. I am 100% sure that no one in the group knew that you were refactoring code when you were multiple times throughout the semester.

Your refactors are very good; you have a good grip and intuition on code modularity. All you need to do is tell people what you think and what you plan on doing.
